### DISCUSSION TITLE
Diffusion Controlled Evaporation with Displacement / Node Deactivation Approach

### DISCUSSION THREAD

**Discussion by AndrewFalkowski** [idaholab/moose#21443] [Q&A General] [answered]
I am looking to model a diffusion controlled evaporation process with mass loss resulting in uniform material recession at the surface; however, I am struggling to come up with a good approach for modeling the recession and shifting of the boundary position with material loss? A coupled displacement condition seems overly complicated and might create errors arising from the displaced mesh. Alternatively, I was considering node deactivation, but I am not sure how to then shift the applied boundary condition.

If I could get some advice on an approach to this, that would be great!

**Comment by GiudGiud** [comment]
@cticenhour @aeslaughter  I think would have had a similar challenge with additive manufacturing (boundary advancing instead of receding)

**Reply by cticenhour** [reply]
I've not done any manufacturing modeling where I've manipulated the mesh, but @SudiptaBiswas and @dewenyushu has done work in additive manufacturing modeling specifically.

**Comment by dewenyushu** [comment, marked as answer]
This looks like something that can be done with the element subdomain modifier, e.g., [`CoupledVarThresholdElementSubdomainModifier`](https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html)

To use it, you will need to define two subdomains in your input. One is the 'active' subdomain containing the elements that belongs to the fluid ('active') material. The other one is the 'dummy' subdomain which contains the elements that belongs to the gas ('loss') material. 

Meanwhile, if you do not solve anything for the evaporated phase, the `kernels` and `materials` blocks need to be block-restricted, i.e., defined only in the 'active' subdomain.

The evaporation of material can be modeled by moving the elements from the 'active' subdomain to the 'dummy' subdomain. This can be taken care of by the `CoupledVarThresholdElementSubdomainModifier`. You will need to couple it with the variable that represents the material loss (e.g., density) and specify a threshold value, below which the element is treated as belonging to the loss material thus can be moved to a 'dummy' subdomain. If you specify the `moving_boundary_name`, in the `CoupledVarThresholdElementSubdomainModifier`, the boundary between the fluid and gas phases should be updated correspondingly.

**Reply by AndrewFalkowski** [reply]
Thank you, this was very helpful! If I am not interested in solving for the gas phase, what would be the best approach for nullifying the influence of variables now in the dummy blocks?

**Reply by dewenyushu** [reply]
You should be able to use the `NullKernel` for the dummy blocks for this purpose. Make sure that you have all your kernels properly defined in corresponding subdomains by including the `block = '<block_ids>'` in every kernel.

**Reply by GiudGiud** [reply]
Either that or you block-restrict the variables out of the dummy blocks.

**Reply by GiudGiud** [reply]
this later approach will be cheaper because the matrix will be smaller.

**Reply by dewenyushu** [reply]
Yeah, I agree with @GiudGiud  - just to add that you may need to turn off `kernel_coverage_check` and/or `material_coverage_check` if you go with the later.

**Reply by AndrewFalkowski** [reply]
Block restriction seems like the right move. 

I have also noticed that the defined boundary seems to persist at the original location even when the `moving_boundary_name` option is supplied. The result, when opened in paraview, shows two boundaries of the same name, one at the original location and the other at the dummy domain interface. Furthermore, it appears that boundary condition property at the dummy domain interface inherits from the persistent one - i.e. the flux at the interface falls to zero. I couldn't find an example in the [MOOSE test files](https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html) that moves a pre-existing boundary (only one that creates a new one), so I am unsure if this is a bug or an issue with my implementation.

**Reply by dewenyushu** [reply]
Glad that it works to some degree.

Would you mind sharing some screen shots to illustrate the issue?

**Reply by AndrewFalkowski** [reply]
Yes, after poking around more it seems the issue is the the boundary condition no longer applies to the new boundary defined by the user object. Here I am tracking the boundary condition value over time and note that the flux (defined by simple Neumann condition) falls to zero once the boundary is moved - creating a natural (zero flux) boundary condition at the new, moved, boundary. I suspect that the original boundary condition is set to a quadrature point and needs to be updated with the moving boundary.  

![surface_flux](https://user-images.githubusercontent.com/3618750/178795653-62e61f11-29ac-472d-ba10-6bc8aecb5a92.png)

I have included a simplified input file that illustrates the behavior below:
```
[Mesh]
  [geo]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 100
    ny = 100
    xmin = 0
    xmax = 1
    ymin = 0
    ymax = 1
  []
  [material_domain]
    input = geo
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    block_id = 1
    block_name = material
    top_right = '0.5 1 0'
  []
  [void_domain]
    input = material_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0.5 0 0'
    block_id = 2
    block_name = void
    top_right = '1 1 0'
  []
  [interface]
    type = SideSetsBetweenSubdomainsGenerator
    input = void_domain
    primary_block = 1
    paired_block = 2
    new_boundary = 'phase_interface'
  []
[]
[UserObjects]
  [material_recession] # object controlling the subdomain shift
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = phi
    block = 1
    criterion_type = BELOW
    threshold = 9
    subdomain_id = 2
    moving_boundary_name = 'phase_interface'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]
[Variables]
  [phi]
    block = 'material'
  []
[]
[AuxVariables]
  [flux_out]
    block = 'material'
  []
[]
[Kernels]
  [mass_diffusion]
    type = ADMatDiffusion
    variable = phi
    diffusivity = 1e-5
    block = 'material'
  []
  [mass_diffusion_time]
    type = ADTimeDerivative
    variable = phi
    block = 'material'
  []
[]
[BCs]
  [flux]
    type = NeumannBC
    boundary = 'phase_interface'
    variable = phi
    value = -1.5e-3
    save_in = flux_out
  []
[]
[ICs]
  [init_phi]
    type = ConstantIC
    block = 'material'
    value = 10
    variable = phi
  []
[]

[Problem]
  coord_type = XYZ
  kernel_coverage_check = false
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  scheme = 'crank-nicolson'
  start_time = 0.0
  end_time = 30
  dt = 1
[]

[Postprocessors]
  [surface_flux]
    type = ElementExtremeValue
    variable = flux_out
    block = 'material'
  []
[]

[Outputs]
  [exo]
    type = Exodus
    output_material_properties = false
    file_base = 'problems/outputs/flux_experiment'
  []
[]

```

**Reply by hugary1995** [reply]
Yeah, I pulled down your example, and I can confirm that the NeumannBC is still looping over the original sideset after the phase_interface has been updated.

BTW, why is this marked as answered, again?

**Reply by hugary1995** [reply]
Okay, I think I know what's going on. We didn't add the neighbor side to the boundary info. I'll have a PR up momentarily.

**Reply by jain651** [reply]
I had the same issue. Do we want to add neighbor side to the moving boundary or we want to create a new boundary (say complement_moving_boundary). I am not sure about either. Please have a look at this: #21383

**Reply by hugary1995** [reply]
This is your input file with the proper fix. Let me know if this is what you were expecting.

![demo](https://user-images.githubusercontent.com/31555204/178824584-08edef35-711e-4651-9aff-a42cb35cfa67.gif)

**Reply by AndrewFalkowski** [reply]
Yes, that looks to be the correct behavior.

**Reply by hugary1995** [reply]
@jain651 I don't quite understand where you want the complement moving boundary to be. Maybe a picture will be helpful. I can talk to you or Ben.

**Reply by jain651** [reply]
The `complement_moving_boundary` would be a collection of newly created neighbor sides, i.e., on the unmoved subdomain. If we want to apply `BCs` to the unmoved domain, we would use `complement_moving_boundary` else to apply `BCs` to moved domain we would use `moving_boundary`. The `complement_moving_boundary` will be an additional feature. However, I cannot think of a practical use of `complement_moving_boundary` now, but there could be one.

**Reply by hugary1995** [reply]
FYI, #21580 is up.

**Reply by hugary1995** [reply]
> The `complement_moving_boundary` would be a collection of newly created neighbor sides, i.e., on the unmoved subdomain. If we want to apply `BCs` to the unmoved domain, we would use `complement_moving_boundary` else to apply `BCs` to moved domain we would use `moving_boundary`. The `complement_moving_boundary` will be an additional feature. However, I cannot think of a practical use of `complement_moving_boundary` now, but there could be one.

I agree that could potentially useful, but that sounds like a different feature than what we are talking about in this discussion.

**Reply by hugary1995** [reply]
@jain651 I think I understand what you meant now. I'll think about how we might accomplish that.
